realm/realm-library/src/main/java/io/realm/internal/RealmNotifier.java

```diff
@@ -97,10 +98,15 @@ protected RealmNotifier(SharedRealm sharedRealm) {
     // called from java_binding_context.cpp
     void didChange() {
         realmObserverPairs.foreach(onChangeCallBack);
-        for (Runnable runnable : transactionCallbacks) {
+
+        Iterator<Runnable> it = transactionCallbacks.iterator();
+        while (it.hasNext()) {
+            Runnable runnable = it.next();
+            // The callback needs to be removed from list before calling to avoid sync transactions in the callback
+            // triggers it recursively.
+            it.remove();
```

Comment:
This reminds me of the good old times of

    while(!transactionCallbacks.isEmpty()) {
         Runnable runnable = transactionCallbacks.remove(0);
         runnable.run();
    }